### ISSUE TITLE
在高通HTP上运行QwenVL2.5

### ISSUE THREAD

**Issue by hcqylymzc** [DakeQQ/Native-LLM-for-Android#36]
大佬tql！
请教一下我想用QNN SDK或者genie sdk来运行VLM，但是目前大佬的代码是输出的动态形状，如果静态形状那LLM部分该怎么处理呢，尤其KV-Cache部分，大佬有什么建议吗？

**Comment by DakeQQ** [comment]
根据我们了解到的情况，目前onnxruntime 1.22.0 + QNN SDK 2.34 仍然无法处理动态尺寸的模型。我们也还在等待onnxruntime更新适配。我们暂未清楚onnxruntime-GenAI使用了啥技术手段克服动态尺寸问题，但从模型来源来看，猜测是做了原生QNN+整型量化的调用接口。但这接口目前无法在一般onnxruntime是取得。

**Comment by DakeQQ** [comment]
我们尝试过了，结论是效果不佳。事实上，可以利用数学手段将所有必要的计算以静态尺寸方式输出。然而，这会频繁的在CPU-共享内存-HTP之间频繁交互数据，如此操作理论上会慢，试验证明亦如此。此外，如果不进行整型量化，那onnxruntime-QNN-HTP只能使用纯FP16的格式。我们测试在0.5B-FP16，解码速度仅仅只有5 token/s。

**Comment by hcqylymzc** [comment]
> 不是我不分享...是太烂就删除了，代碼已不可考。但能分享一下关键经验，就是动态尺寸主要是受到了`attention`里面的`softmax`算子影响，才比须使用动态。毕竟，transformer模型的原理是根据上一个字接龙下一个字的，每次计算都是固定的尺寸，但卡了个softmax （对过往已吐的字做logit计算，才必须随着字数动态扩增）
> 
> Prefill阶段亦如此，表现上看是一次性计算所有input_ids，但attention_mask起到了遮掩作用，本质上仍然可以用一条固定尺寸的张量逐步计算完成。

非常感谢大佬的解答~我再尝试一下看看

**Comment by hcqylymzc** [comment]
> 不是我不分享...是太烂就删除了，代碼已不可考。但能分享一下关键经验，就是动态尺寸主要是受到了`attention`里面的`softmax`算子影响，才比须使用动态。毕竟，transformer模型的原理是根据上一个字接龙下一个字的，每次计算都是固定的尺寸，但卡了个softmax （对过往已吐的字做logit计算，才必须随着字数动态扩增）
> 
> Prefill阶段亦如此，表现上看是一次性计算所有input_ids，但attention_mask起到了遮掩作用，本质上仍然可以用一条固定尺寸的张量逐步计算完成。

大佬我基于你的代码导出了静态图的onnx在htp上运行3B模型，发现奇慢无比1token/s都没有，有什么经验可以借鉴吗，算子合并，优化之类的。

**Comment by DakeQQ** [comment]
我也不清楚您改的模型都变啥结构了，因此现在也没有头绪。您先修改打印设定，看看ONNXRuntime日志里面的算子分布情况，查查有多少分给了QNN-HTP,  多少在CPU上。我猜测您也遇到了`频繁在CPU-共享内存-HTP之间交互通信`的情况。
```cpp
ort_runtime_A->CreateEnv(ORT_LOGGING_LEVEL_VERBOSE, "myapplication", &ort_env_A);
```
